### DISCUSSION TITLE
JMESPath Community Features 1.2 Brainstorming

### DISCUSSION THREAD

**Discussion by springcomp** [jmespath-community/jmespath.spec#151] [Announcements]
This discussion has been started to list topics that could be included in a next version of JMESPath. This is a good place to discuss proposals and new features.

Features with a ✅ check-mark are available or prototyped in most JMESPath Community implementations already.
Features with a 📃 symbol have a reasonable first JEP draft. Please, share your feedback.

<img src="https://user-images.githubusercontent.com/1768820/169634102-039f2b99-04fb-4433-b683-e2bbe07a9cc3.png" height="50" />

Here are the topics we currently have:

- Structural Manipulations
  - [Recursive Tree Traversal](https://github.com/jmespath-community/jmespath.spec/discussions/148#discussioncomment-13135324) 📃
- Functional Programming
  - Obsolete JEP-11 in favor of a newly proposed [`let-expression`](https://github.com/jmespath/jmespath.jep/pull/18). ✅
  - **JEP-11a** [Lexical scope](https://github.com/jmespath-community/jmespath.spec/blob/main/jep-011a-lexical-scope.md) ✅
  - [reduce](https://github.com/jmespath/jmespath.jep/issues/25) function
- Functions
  - The [`date()`](https://github.com/jmespath-community/jmespath.spec/discussions/169) function.
  - The [`indices()` and `range()`](https://github.com/jmespath-community/jmespath.spec/discussions/49#discussioncomment-6258917) functions. 📃
- Miscellaneous Topics
  - [Ordering Arrays and Strings](https://github.com/jmespath-community/jmespath.spec/discussions/91#discussioncomment-6258367). 📃
  - [Third-Party Functions](https://github.com/jmespath-community/jmespath.jep/discussions/20) 📃
  - **JEP-20** [Compact-style for multi-select-hash](https://github.com/jmespath-community/jmespath.spec/discussions/19#discussion-3913995) 📃
- **JEP-21** [Ternary Conditional Expression](https://github.com/jmespath-community/jmespath.spec/discussions/179) ✅

Please, feel free to to advocate for yet [more of the features we had been brainstorming](https://github.com/jmespath-community/jmespath.spec/discussions/38) last year but did not make it to the current release.

## History

|Date|Description|
|---|---|
|2022-11-03|Compact Syntax for `multi-select-hash` proposal has been assigned number JEP-020
|2023-07-10|[Proposal for `match` and `like`](https://github.com/jmespath/jmespath.jep/issues/29) contextual keywords or functions for comparisons is postponed until [spec](https://www.ietf.org/archive/id/draft-ietf-jsonpath-iregexp-08.html) reaches RFC-status, if ever.
|2025-15-01|Linked to [Ternary Operator Expression](https://github.com/jmespath-community/jmespath.spec/discussions/179) proposal draft
|2025-05-14|Ternary Operator proposal has been assigned number JEP-021

**Comment by FilipPyrek** [comment]
The `date()` function is a good idea 👍 👍

**Comment by Amerousful** [comment]
Recursive tree traversal is undeniably a valuable feature. Really looking forward to it.
